### ISSUE TITLE
Updating packages on pkg.mxe.cc / using Debian Stretch

### ISSUE THREAD

**Issue by ooxi** [mxe/mxe#2021]
Would it be possible to update the prebuild packages available on pkg.mxe.cc? Also I'm wondering whether or not Debian Stretch could be supported, too.

:+1: for the great work so far!

**Comment by ooxi** [comment]
@starius is there maybe a way I can help with this e.g. by providing compiled binaries?

**Comment by mlocati** [comment]
Also: what about automatically updating pkg.mxe.cc when pushing to master on mxe? This could be done in a TravisCI job...

**Comment by ooxi** [comment]
> Also: what about automatically updating pkg.mxe.cc when pushing to master on mxe? This could be done in a TravisCI job...

That's not trivial as a full build takes about two days on a _really_ beefy machine :-)

**Comment by tonytheodore** [comment]
And that beefy machine tends to be your main machine so you can't lock it up for that long!

@mlocati The main reason we don't incrementally update binary packages is that they're supposed to be a stable working set that has been tested together.

@ooxi I believe the packages need to be built on a Debian Wheezy machine as the version of glibc needs to be a lowest common denominator across the various CI platforms and Debian derivatives. (At least that's my understanding of why we needed to extend support for Wheezy, it could be something else now). That likely means it's a cloud server or local vm - the cloud server takes even longer to build and the local vm doesn't have the bandwidth to upload packages...

@starius I'll do a full `build-pkg` test of #1990 after #2057, I imagine it doesn't matter how long it takes if there's a known working commit.

**Comment by tonytheodore** [comment]
Some packages are broken on Wheezy (build still running), notably qt 5.10 which seems to need a newer c++ compiler for the native qmake etc. I wonder if we could use Ubuntu 14.04 instead?

**Comment by ooxi** [comment]
I think Travis CI is using Ubuntu 14.04 by default, thus a majority of our users would benefit directly from using it as a base.

**Comment by tonytheodore** [comment]
`build-pkg` eventually fails with `git merge` issues caused by `waf` generating artefacts at runtime - probably needs a `.gitignore`.

**Comment by tonytheodore** [comment]
> build-pkg eventually fails with git merge issues

These aren't actually failures, just warnings that happen to be at the end of the build. They're present in the last build-pkg run and don't seem to affect anything.

> probably needs a .gitignore

Strangely, neither `.gitignore` or `$GIT_DIR/info/exclude` work, so I've just [used `waf` in the build dir instead of installing generally](https://github.com/mxe/mxe/pull/2078/commits/dd39411f038e96985ff4a232b22e4dd37f3b9d55)

**Comment by darealshinji** [comment]
Would like the packages to be updated too. I'm using them to compile small programs on Bash on Windows, and bootstrapping MXE didn't work very well in past.

**Comment by ooxi** [comment]
@darealshinji fyi it's now possible to build only a subset of packages [using `MXE_BUILD_PKG_PKGS`](https://github.com/mxe/mxe/blob/master/tools/build-pkg.lua#L39). Maybe this helps with bootstrapping :-)

**Comment by tonytheodore** [comment]
I've started a test build with Debian Stretch - should take around 40 hours.

**Comment by tonytheodore** [comment]
Ended up taking a very long time (maybe 100 hours) but there are some things to fix: https://gist.github.com/tonytheodore/73a421b81a8c4944e72a1f9383e30380

**Comment by ooxi** [comment]
Should we maybe remove those packages? If nobody noticed, they are broken, maybe nobody cares?

**Comment by darealshinji** [comment]
Suggestion: maybe a "lite" version that contains only libraries that are used often and take long time to compile (Qt, Gtk, boost, etc.) could be updated more often, like every 3 months.

**Comment by ashkulz** [comment]
I have a server at Kimsufi, which I can contribute for the rebuild every so often. I'm willing to take over maintenance of this task if I get some input:

- I assume https://github.com/mxe/mxe-apt is to be used? There doesn't seem to be any support for stretch.
- Any idea of how to sign with alternate GPG key? Is the current key still valid?

I'm willing to update the docs/wiki if I get some pointers :smile:

**Comment by ooxi** [comment]
That would be gread @ashkulz! https://github.com/mxe/mxe-apt is integrated into this repository, no need for an external tool.

In order to build all packages for Debian you have to invoke [build-pkg.lua](https://github.com/mxe/mxe/blob/master/tools/build-pkg.lua). See the included documentation for details.

If you have any questions regarding setup, I'll be glad to help out. However, I don't know abut signing with GPG, somebody else has to step in here :-)

**Comment by ashkulz** [comment]
I've created `tools/Dockerfile.stretch` for performing the build in a clean environment:

```
FROM debian:stretch

RUN apt-get update && apt-get install -y -q --no-install-recommends \
    autoconf \
    automake \
    autopoint \
    bash \
    bison \
    bzip2 \
    dpkg \
    dpkg-dev \
    flex \
    g++ \
    g++-multilib \
    gettext \
    git \
    gperf \
    fakeroot \
    intltool \
    libc6-dev-i386 \
    libgdk-pixbuf2.0-dev \
    libltdl-dev \
    libssl-dev \
    libtool-bin \
    libxml-parser-perl \
    lua5.1 \
    make \
    openssl \
    p7zip-full \
    patch \
    perl \
    pkg-config \
    python \
    ruby \
    scons \
    sed \
    unzip \
    wget \
    xz-utils \
    && rm -rf /var/lib/apt/lists/*
```

and have triggered off the build with the following commands:

```bash
docker build -t mxe-rebuild -f tools/Dockerfile.stretch tools/
docker run --rm --user `id -u`:`id -g` -v$PWD:/usr/lib/mxe -w/usr/lib/mxe mxe-rebuild lua tools/build-pkg.lua
docker rmi mxe-rebuild
```
It's probably going to take a long time, so I'll post an update when it is done.

**Comment by ashkulz** [comment]
Unfortunately the build hung after the first pass, so not really sure what's up with that. No .deb packages were created either, even after I left it running for 12 hours in the hope it would move forward. Here's [the log](https://github.com/mxe/mxe/files/2101931/mxe.log) of the build, which has some failures and merge conflicts at the end. I'm re-running the build and seeing if there's a different result.

**Comment by ashkulz** [comment]
So pass 1 succeeded -- I used a trusty environment as it easier to use via Travis CI, as discussed above. The docker image was created using `tools/Dockerfile.trusty`:
```
FROM debian:stretch

RUN apt-get update && apt-get install -y -q --no-install-recommends \
    autoconf \
    automake \
    autopoint \
    bash \
    bison \
    bzip2 \
    dpkg \
    dpkg-dev \
    flex \
    g++ \
    g++-multilib \
    gettext \
    git \
    gperf \
    fakeroot \
    intltool \
    libc6-dev-i386 \
    libgdk-pixbuf2.0-dev \
    libltdl-dev \
    libssl-dev \
    libtool \
    libxml-parser-perl \
    lua5.1 \
    make \
    openssl \
    p7zip-full \
    patch \
    perl \
    pkg-config \
    python \
    ruby \
    scons \
    sed \
    unzip \
    wget \
    xz-utils \
    && rm -rf /var/lib/apt/lists/*
```
The commands run to make the build were:
```
docker build -t mxe-trusty -f tools/Dockerfile.trusty tools/
docker run --rm -v$PWD:/usr/lib/mxe -w/usr/lib/mxe -e MXE_BUILD_PKG_NO_FAKEROOT=1 mxe-trusty lua tools/build-pkg.lua
docker rmi mxe-trusty
```
which builds on top of #2153 to skip using fakeroot. The packages are available (with the complete logs) from https://us.wkhtmltopdf.org/mxe/ -- the second pass is still running, and I've yet to figure out how to use reprepro with my GPG key.

**Comment by ashkulz** [comment]
Can some of the core contributors step in and give feedback on:
- whether PRs for getting this working in Docker would be acceptable
- which distributions to target (or use something like alpine so that it works on any distribution due to static linkage, but will increase the size of host packages)
- you are OK with me contributing full MXE builds on a regular basis (monthly/biweekly)
- how to handle the errors in the full rebuild

**Comment by tonytheodore** [comment]
Apologies for the late reply, I don't know much about apt, gpg, docker etc.

I've made some tweaks to build-pkg (#2256) that will allow [multiple dists](https://mirror.mxe.cc/repos/apt/dists/) and should make it fairly easy to accept contributed builds and lite/custom variants (signing and disk space aside).

@ashkulz PRs for building in docker and contributed builds would be welcome if you still have the time! Now that I know a little more, creation of debs can happen anywhere so you only need the gpg setup on one machine.

I think we should target Travis CI default distro and Debian stable to start with. I've added a `codename` variable that should let us do lowest-common-glibc based versions down the track.

For dealing with errors, regular full builds should help get them fixed sooner but we can also look at breaking `build-pkg` into a more incremental process.

**Comment by ooxi** [comment]
Thank you very much for your work @tonytheodore! I don't quite get the tweak to build-pkg, are we now having multiple packages for different dists or are we targeting the lowest common denominator?

**Comment by tonytheodore** [comment]
@ooxi wrote:
> are we now having multiple packages for different dists

Yes, multiple packages built on the target distro. For now that's trusty, xenial, and stretch but I think we can do Ubuntu LTS and Debian Stable & Testing (keeping old stable around for a transition time).

 > or are we targeting the lowest common denominator

Down the track we can do something like a lowest common denominator with tarballs only for distros like Arch and Void that keep updating very quickly - this will be something more like "most recent common libc".

**Comment by ooxi** [comment]
@tonytheodore 

> Yes, multiple packages built on the target distro. For now that's trusty, xenial, and stretch but I think we can do Ubuntu LTS and Debian Stable & Testing (keeping old stable around for a transition time).

I really like this approach, since it's much more scaleable!

**Comment by ashkulz** [comment]
> PRs for building in docker and contributed builds would be welcome if you still have the time! Now that I know a little more, creation of debs can happen anywhere so you only need the gpg setup on one machine.

I've started work on this, currently rebuilding with `xenial` which should take ~60 hours or so (20 hours already done). My goals are:

1. **Have up-to-date builds**: I can run a cron job every day which rebuilds everything. Considering that `ccache` is being used, it hopefully won't take too much time. I plan to only support `xenial` and `stretch`, as `trusty` will be EOL this month-end.
2. **Include feedback for PRs**: Ideally, once a PR is proposed we can automatically run a build on that branch so that we can figure out whether it compiles.

Not sure I'll get around to part 2 soon, but 1 is doable. Short term roadmap looks like:
- send PR for Docker images
- create `tools/build-repo.lua` which uses the Docker images and creates a signed apt repository and send PR
- publish the apt repository for testing purposes
- get the rebuild to occur quickly and downstream packages should also be rebuilt automatically (if not, figure out some way to use `show-downstream-deps` to do that) -- the "incremental" part @tonytheodore talked above
- setup cron to do this at a given interval

I also think it'd be great if we separate out the daily repo updates with some sort of snapshot: e.g. once every 2 months. That can be on `pkg.mxe.cc` as now while this can be on another directory which is rsynced to **or** have `latest-pkg.mxe.cc` as an alias for that server (signed with a different key).

**Comment by ashkulz** [comment]
So, the results of the first pass are available at http://us.wkhtmltopdf.org/mxe/xenial/ -- the following packages failed to build:
```
x86_64-w64-mingw32.shared~nsis
x86_64-w64-mingw32.shared~libwebsockets
x86_64-w64-mingw32.static~libwebsockets
x86_64-w64-mingw32.static~nsis
x86_64-w64-mingw32.static~gst-plugins-bad
x86_64-w64-mingw32.shared~gst-plugins-bad
i686-w64-mingw32.static~nsis
i686-w64-mingw32.static~libwebsockets
i686-w64-mingw32.static~gst-plugins-bad
i686-w64-mingw32.shared~nsis
i686-w64-mingw32.shared~itk
i686-w64-mingw32.shared~libwebsockets
i686-w64-mingw32.shared~gst-plugins-bad
```
@tonytheodore, is the second pass really required? Would also appreciate feedback on the above proposal.

**Comment by tonytheodore** [comment]
Thanks for looking at this @ashkulz, the second pass isn't necessary when building deb packages - it's more a QA test that should be run previously to identify dependency and ordering issues. All packaging builds should set `MXE_BUILD_PKG_NO_SECOND_PASS=1`.

> I also think it'd be great if we separate out the daily repo updates with some sort of snapshot

I [recently added a version id](https://github.com/mxe/mxe/commit/1fdeda5c95da936a46272a51d8fa02dca9520997) so tagged builds should additionally set `MXE_BUILD_PKG_VERSION_ID=$tag`, where `$tag` should ideally be a [complete build](https://github.com/mxe/mxe/releases/tag/build-2019-03-04) (in this case a fix for itk i686-w64-mingw32.shared eludes me). I was thinking every three months, but it may end up being two or six months depending on how much users want a "stable" channel.

I was leaning towards the docker approach and will have a new (beefy) machine shortly that I can use instead of spinning up cloud vms for different distros. 

I agree with the general approach of your proposal, I probably wouldn't use lua and am not sure about multiple signing keys - but getting the infrastructure in place is the most important part. My end goal would be using pre-built packages to bootstrap a custom repo.

**Comment by ashkulz** [comment]
> All packaging builds should set `MXE_BUILD_PKG_NO_SECOND_PASS=1`.

Will do!

> I was thinking every three months, but it may end up being two or six months depending on how much users want a "stable" channel.

I don't think it makes sense to wait so long, think about security fixes e.g. OpenSSL. If we've got users for the pre-built packages, they probably want to use the latest version else they're going to freeze the MXE commit to use. So I'm in favor of a short cycle: daily and monthly -- else we'll end up requiring too much disk space.

> I was leaning towards the docker approach and will have a new (beefy) machine shortly that I can use instead of spinning up cloud vms for different distros.

As I said earlier, I have a dedicated server (not cloud VMs) for development. Unfortunately, the build is not parallelized so it doesn't use the 24 threads at all, load average is <= 4 for most of the build. I did look into parallelization but doesn't seem to be easily feasible with the current approach.

> I agree with the general approach of your proposal, I probably wouldn't use lua and am not sure about multiple signing keys - but getting the infrastructure in place is the most important part. My end goal would be using pre-built packages to bootstrap a custom repo.

I'd be happy to do this in Python :smile: I didn't get the part about bootstrapping a custom repo, can you elaborate?

**Comment by tonytheodore** [comment]
Bootstrapping a custom repo is really just this:
  - use binary packages to create mxe source installation
  - modify mxe source for selected packages
  - build new binary packages and publish to repo

The "modify" is pretty broad, it could be pinning versions, fixes, package updates etc. For example, it should be easy to install the results from http://us.wkhtmltopdf.org/mxe/xenial/, fix the broken packages, and upload the delta.

It's really just iteratively switching between source and binary installs.
